### ISSUE TITLE
Null ref in DictionarySlim resize

### ISSUE THREAD

**Issue by Zoxive** [Zoxive/Apex.Runtime#2]
```
System.NullReferenceException: Object reference not set to an instance of an object.\r\n
   at Xamarin.Apex.Runtime.Internal.DictionarySlim`2.Resize()\r\n
   at Xamarin.Apex.Runtime.Internal.DictionarySlim`2.AddKey(TKey key, Int32 bucketIndex)\r\n
   at Xamarin.Apex.Runtime.Internal.DictionarySlim`2.GetOrAddValueRef(TKey key)\r\n
   at Xamarin.Apex.Runtime.Memory.HasNoReferences(Type elementType)\r\n
   at System.Linq.Enumerable.All[TSource](IEnumerable`1 source, Func`2 predicate)\r\n
   at Xamarin.Apex.Runtime.Memory.HasNoReferences(Type elementType)\r\n
   at Xamarin.Apex.Runtime.Mode.CompiledExpressionsMode.DynamicCode.GenerateMethodImpl(Type type, Boolean isVirtual)\r\n
   at Xamarin.Apex.Runtime.Mode.CompiledExpressionsMode.GetSizeOfSealedInternal[T](T obj, Memory memory)\r\n
   at Apex.Runtime.CompiledExpression_SizeOf_Engine.Stateless.Rules.VariableRule(Closure , Object , Memory )\r\n
   at Xamarin.Apex.Runtime.Mode.CompiledExpressionsMode.GetSizeOfInternal[T](T obj, Memory memory)\r\n
   at Apex.Runtime.CompiledExpression_SizeOf_Engine.Stateless.RuleBase[](Closure , RuleBase[] , Memory )\r\n
   at Apex.Runtime.CompiledExpression_SizeOf_System.Collections.Generic.List`1[[Engine.Stateless.RuleBase, Engine.Stateless, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]](Closure , Object , Memory )\r\n
   at Xamarin.Apex.Runtime.Mode.CompiledExpressionsMode.GetSizeOfInternal[T](T obj, Memory memory)\r\n
   at Apex.Runtime.CompiledExpression_SizeOf_Engine.Stateless.Rules.ConditionRule(Closure , Object , Memory )\r\n
   at Xamarin.Apex.Runtime.Mode.CompiledExpressionsMode.GetSizeOfInternal[T](T obj, Memory memory)\r\n
   at Apex.Runtime.CompiledExpression_SizeOf_Engine.Stateless.Rules.ConditionRule(Closure , Object , Memory )\r\n
   at Xamarin.Apex.Runtime.Mode.CompiledExpressionsMode.GetSizeOfInternal[T](T obj, Memory memory)\r\n
   at Apex.Runtime.CompiledExpression_SizeOf_Engine.Stateless.RuleBase[](Closure , RuleBase[] , Memory )\r\n
   at Apex.Runtime.CompiledExpression_SizeOf_System.Collections.Generic.List`1[[Engine.Stateless.RuleBase, Engine.Stateless, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]](Closure , Object , Memory )\r\n
   at Xamarin.Apex.Runtime.Mode.CompiledExpressionsMode.GetSizeOfInternal[T](T obj, Memory memory)\r\n
   at Apex.Runtime.CompiledExpression_SizeOf_Engine.Stateless.Rules.VirtualRootRule(Closure , Object , Memory )\r\n
   at Xamarin.Apex.Runtime.Mode.CompiledExpressionsMode.GetSizeOfInternal[T](T obj, Memory memory)\r\n
   at Apex.Runtime.CompiledExpression_SizeOf_Engine.Stateless.Models.Ruleset(Closure , Ruleset , Memory )\r\n
   at Xamarin.Apex.Runtime.Memory.SizeOf[T](T obj)\r\n
   at Engine.Caching.SizeOfObject.SizeOf[T](T obj) in C:\\Build01\\_work\\61\\s\\src\\SuperEngine\\Engine.Caching\\MemorySizeComputePool.cs:line 50\r\n
   at Zoxive.Memoized.Memoized_RulesetRepository.FindAsync(Specification __iRuleSetSpec) in C:\\Build01\\_work\\61\\s\\src\\SuperEngine\\Engine.Stateless.MultiTenant\\Zoxive.MemoizeSourceGenerator\\Zoxive.MemoizeSourceGenerator.Generator\\Memoized_RulesetRepository.cs:line 45\r\n
   at Engine.Stateless.Rules.CreateComponentRule.Execution.GetRulesetAndSpecAsync(CreateComponentRule rule, Value root, IEngineExpressionEvaluator expressionEvaluator, IRulesetRepository rulesetRepository, Component activeComponent, String profile) in C:\\Build01\\_work\\61\\s\\src\\SuperEngine\\Engine.Stateless\\Rules\\CreateComponentRule.cs:line 113\r\n
   at Engine.Stateless.Rules.CreateComponentRule.Execution.ExecuteRootAsync(CreateComponentRule rule, Value root, ImmutableArray`1 rapidOptions, IEngineExpressionEvaluator expressionEvaluator, IRulesetRepository rulesetRepository, IComponentRulesetUtility componentRulesetUtility, ImmutableSortedDictionary`2 activeComponents, ImmutableSortedDictionary`2 inactiveComponents, String profile) in C:\\Build01\\_work\\61\\s\\src\\SuperEngine\\Engine.Stateless\\Rules\\CreateComponentRule.cs:line 83\r\n
   at Engine.Stateless.RuleProcessor.ExecuteCreateRootComponentRuleAsync(CreateComponentRule createComponentRule, ExecutionState state) in C:\\Build01\\_work\\61\\s\\src\\SuperEngine\\Engine.Stateless\\RuleProcessor.Executes.cs:line 103\r\n
   at Engine.Stateless.RuleProcessor.ExecuteCurrentRuleAsync(ExecutionStateRoot state) in C:\\Build01\\_work\\61\\s\\src\\SuperEngine\\Engine.Stateless\\RuleProcessor.cs:line 322\r\n
   at Engine.Stateless.RuleProcessor.ExecuteAsync() in C:\\Build01\\_work\\61\\s\\src\\SuperEngine\\Engine.Stateless\\RuleProcessor.cs:line 72\r\n
   at Engine.Stateless.RuleEngine.ExecuteAsync() in C:\\Build01\\_work\\61\\s\\src\\SuperEngine\\Engine.Stateless\\RuleEngine.cs:line 158\r\n
   at Engine.Stateless.RuleEngine.ExecuteToCompletionAsync() in C:\\Build01\\_work\\61\\s\\src\\SuperEngine\\Engine.Stateless\\RuleEngine.cs:line 205
```

**Comment by Zoxive** [comment]
Fixed with #3

**Comment by Zoxive** [comment]
If the issue still happens its most likely because you are using the same `Memory` instance in multiple threads. `Memory` is not threadsafe so use ObjectPooling.
